pkg/reconciler/revision/background.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package revision
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/authn/k8schain"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/workqueue"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// imageResolver is an interface used mostly to mock digestResolver for tests.
+type imageResolver interface {
```

Comment:
cc @ImJasonH @vdemeester @jonjohnsonjr @bobcatfish @dlorenc 

I'm fairly certain Tekton has the same problem that we do here (since you rewrite container entrypoints).  I suspect that we could generalize this approach to cache and return more of the image metadata and move it to PKG, if that would be useful.

Reply:
If I understand correctly, instead of blocking in the reconciler, this queues up digest resolution to a worker pool, updates revision status to `ResolvingDigests`, then re-enqueues the revision when digest resolution completes? Smart 👍. This hits all the stuff I used to be interested in.

I wonder if it would make sense to just make image/digest an actual child resource of revision? I'm not sure if that would buy you anything but half of this stuff seems like it's re-implementing the kube informer machinery.

>  I suspect that we could generalize this approach to cache and return more of the image metadata and move it to PKG, if that would be useful.

Which parts do you want to generalize? Seems like the registry bits aren't really relevant and this would apply equally to any long-running off-cluster operation.

This pattern might also apply to https://github.com/tektoncd/pipeline/pull/3142?

Reply:
I'm definitely _interested_, but I'm a little hesitant to share this with knative/pkg prematurely since we might end up having diverging needs/constraints. I will be keeping an eye on it though 👀 

I haven't instrumented anything but I'd be pretty surprised if digest resolution took a noticeably long amount of time in any real scenario. Does Knative just have much lower latency expectations than Tekton (currently) does?

Reply:
If we can make the entrypoint and cmd available that would be good. I have a need for that info too and forked resolve() to get it as part of the revision metadata.

Reply:
> If we can make the entrypoint and cmd available that would be good

If you want to do this, keep in mind:
1. You'll need to handle [schema 1](https://docs.docker.com/registry/spec/manifest-v2-1/) images differently.
2. You'll need more information to determine the correct platform for [manifest lists](https://docs.docker.com/registry/spec/manifest-v2-2/#manifest-list)

Reply:
> You'll need to handle schema 1 images differently.

Or not at all! _(ducks)_

@ImJasonH @duglin since you have a need for the more generalized form of this, do you want to open an issue to continue the discussion (so we don't hold this up) and we can unblock the actual fix here?

Reply:
Happy to take this out of this review. I'm not convinced Tekton needs to share this _yet_, but I'd like to stay on top of it in case it ends up fitting our needs.